File: frontend/olympic_it_project/olympic_it_project/lib/service/classes_service.dart
import 'package:olympic_it_project/core/api_client.dart';
import 'package:olympic_it_project/core/api_response.dart';
import 'package:olympic_it_project/dto/admin_manager/classes/classes_response.dart';
import 'package:olympic_it_project/dto/admin_manager/classes/create_class_request.dart';
import 'package:olympic_it_project/dto/admin_manager/classes/update_class_request.dart';

class ClassService {
  final _api = ApiClient.instance;

  Future<List<ClassResponse>> getAll({int? academicYearId}) async {
    String url = "admin/classes";

    if (academicYearId != null) {
      url += "?academicYearId=$academicYearId";
    }

    final response = await _api.get(url);

    final apiResponse = decodeApiResponse<List<ClassResponse>>(
      response,
      (data) => (data as List)
          .map((e) => ClassResponse.fromJson(e))
          .toList(),
    );

    if (apiResponse.code == 200 && apiResponse.data != null) {
      return apiResponse.data!;
    }

    throw Exception(apiResponse.message);
  }

  Future<void> create(CreateClassRequest request) async {
    final response = await _api.post(
      "admin/classes",
      request.toJson(),
    );

    final apiResponse = decodeApiResponse<String>(
      response,
      (data) => data?.toString() ?? "",
    );

    if (apiResponse.code != 200) {
      throw Exception(apiResponse.message);
    }
  }

  Future<void> update(int id, UpdateClassRequest request) async {
    final response = await _api.put(
      "admin/classes/$id",
      request.toJson(),
    );

    final apiResponse = decodeApiResponse<String>(
      response,
      (data) => data?.toString() ?? "",
    );

    if (apiResponse.code != 200) {
      throw Exception(apiResponse.message);
    }
  }

  Future<void> delete(int id) async {
    final response = await _api.delete("admin/classes/$id");

    final apiResponse = decodeApiResponse<String>(
      response,
      (data) => data?.toString() ?? "",
    );

    if (apiResponse.code != 200) {
      throw Exception(apiResponse.message);
    }
  }
}

File: admin-web/js/service/classes_service.js
import { apiClient, safeDecode } from "../core/api_client.js";
import { unwrap } from "../core/api_response.js";

const BASE = "admin/classes";

class ClassesService {
    async getAll(academicYearId) {
        const query = academicYearId ? `?academicYearId=${academicYearId}` : "";
        const res = await apiClient.get(`${BASE}${query}`);
        const json = await safeDecode(res);
        return unwrap(json);
    }

    async create(className, academicYearId) {
        const res = await apiClient.post(BASE, { className, academicYearId });
        const json = await safeDecode(res);
        return unwrap(json);
    }

    async update(id, className, academicYearId) {
        const res = await apiClient.put(`${BASE}/${id}`, { className, academicYearId });
        const json = await safeDecode(res);
        return unwrap(json);
    }

    async delete(id) {
        const res = await apiClient.delete(`${BASE}/${id}`);
        const json = await safeDecode(res);
        return unwrap(json);
    }
}

export const classesService = new ClassesService();